### PR TITLE
[inductor] Refactor memory management code in wrapper codegen

### PR DESCRIPTION
**Pull request by desertfire**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #96768

Summary: use inheritance to simplify CppWrapperCodeGen and to prepare for AOT codegen

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10

### CODE REVIEW

**Review by jansel — APPROVED** [review]

File: torch/_inductor/codegen/wrapper.py [hunk 1]
```diff
@@ -213,53 +142,13 @@ def codegen(self, code: IndentedBuffer):
...
-        assert self.reused_as.get_name() not in V.graph.removed_buffers
-        code.writeline(
-            make_buffer_reuse(
+            self.wrapper.make_buffer_reuse(
                 self.node,
                 self.reused_as,
-                del_func=lambda name: f"; {name}.reset()",
-                declare="auto ",
-                ending=";",
-                as_strided="at::as_strided",
```
Comment (ngimel) [inline torch/_inductor/codegen/wrapper.py:243]: do we no longer need to generate cpp `at::as_strided` calls?

Comment (desertfire) [reply torch/_inductor/codegen/wrapper.py:243]: We still need `at::as_strided`, but now it is generated in make_buffer_reuse which works for both WrapperCodeGen and CppWrapperCodeGen.

**Pull request closed without merging**